driver/src/http.rs

```diff
@@ -75,22 +94,165 @@ impl HttpClient {
     }
 
     /// Standard HTTP GET request that parses the result as JSON.
-    pub fn get_json<U, T>(&self, url: U) -> Result<T>
+    fn get_json<U, T>(&self, url: U) -> Result<(T, Duration, usize)>
+    where
+        Uri: TryFrom<U>,
+        <Uri as TryFrom<U>>::Error: Into<HttpError>,
+        T: DeserializeOwned,
+    {
+        let start = Instant::now();
+        let json = self.inner.get(url)?.text()?;
+        let size = json.len();
+        let result = serde_json::from_str(&json)?;
+
+        Ok((result, start.elapsed(), size))
+    }
+
+    /// Standard HTTP GET request that parses the result as JSON.
+    async fn get_json_async<U, T>(&self, url: U) -> Result<(T, Duration, usize)>
+    where
+        Uri: TryFrom<U>,
+        <Uri as TryFrom<U>>::Error: Into<HttpError>,
+        T: DeserializeOwned,
+    {
+        let start = Instant::now();
+        let json = self.inner.get_async(url).await?.text()?;
+        let size = json.len();
+        let result = serde_json::from_str(&json)?;
+
+        Ok((result, start.elapsed(), size))
+    }
+}
+
+/// An HTTP client instance with metrics.
+#[derive(Debug)]
+pub struct HttpClient<Subsystem> {
+    inner: HttpClientInner,
+    metrics: Arc<HttpMetrics>,
+    _subsystem: PhantomData<Subsystem>,
+}
+
+/// Trait for extending HTTP client with labeled version of HTTP methods with
+/// metrics.
+///
+/// This trait allows `HttpClient` to require a label value when a subsystem is
+/// labeled, and to not require a label value when it is unlabeled.
+pub trait LabeledHttpClient<Subsystem> {
```

Comment:
Why do the Labeled and Unlabeled client have different trait signatures? I also don't see where they are being used (I assume it's somehow implicitly derived from whether a metric is labeled or unlabelled)?

Reply:
They are needed so that we can have methods with the same name but different signature:
`http_client.get_json()` and `http_client.get_json(Subsystem::Label)`.